Daemon process

martinyunify opened the pull request:
daemon process  #9 
-- fill nic pool with nic id.
-- maintain nic info table
-- delete cached nic when exit.
-- manage and allocate gateway nic 
-- add version number generator script
-- validate vxnet before launch server #16 
-- update README
-- refactor nic pool 
-- add prometheus monitoring
-- add clean cache command

Commit:
add daemon process

Commit:
remove cleaner
adopt grpc
define protocol
update deps

Commit:
add grpc protocol
add nicpool event loop
refactor provider

Commit:
fix bugs in cni plugin and daemon process.
add log messages

jolestar commented on pkg/server/nicpool.go:
```diff
@@ -0,0 +1,294 @@
+
+	resoureStub *qingcloud.QCNicProvider
+
+	sync.WaitGroup
```
这里 pool 自己本身既是 sync.RWMutex 也是 sync.WaitGroup 看后面的代码的时候感觉会比较奇怪，不如作为独立的字段容易理解。此处仅仅建议。

jolestar commented on pkg/server/nicpool.go:
```diff
@@ -0,0 +1,294 @@
+	go func() {
+		defer pool.Done()
+		//check if nic is in dict
+		pool.RLock()
```
这个  lock 主要是为了保护 pool.nicDict， 我觉得要么扩展一个 sync map 出来，要么用 func 把从 map 中获取数据的操作封装起来，否则这种 lock,unlock 在很多地方出现，容易出错。

jolestar commented on cmd/hostnic/hostnic.go:
```diff
@@ -89,58 +49,57 @@ func cmdAdd(args *skel.CmdArgs) error {
-		return err
-	}
-	nic, err := nicProvider.CreateNic()
+	conn, err := grpc.Dial(conf.BindAddr,grpc.WithInsecure())
```
这个 BindAddr 要有个默认值，如果没配置就用默认值。

jolestar reviewed (commented):
除了上面几个小问题，整体上没大问题，但要注意两点：
1. 通过 pool.Add, pool.Done 进行异步操作的地方较多，如果这块出现 bug 可能导致死锁。
2. NicPool 的逻辑需要想办法通过单元测试验证下，

Commit:
remove dead code
validate vxnet during bootstrap

Commit:
remove dead code
validate

Commit:
gofmt

Commit:
update deps
validate vxnet

Commit:
update deps
validate vxnet

Commit:
fix bugs in provider

Commit:
update readme

Commit:
update readme

Commit:
add deps

add ready pool to recycle nic first

migrate to cocurrent map in gateway and nicinfo

Commit:
fix lifecycle bug
get all items in gateway map

martinyunify commented:
为了避免死锁网卡信息和网卡缓冲池是分开维护的。这里可能有不同步所以加了validationfunc确保拿到的网卡在线。加waitgroup是因为需要在关闭的时候删除返还的网卡。只有在关闭的时候才会调用wait。

Commit:
gofmt

Commit:
refactor to adopt unit test
goimports

Commit:
provide default value for bindaddr in cni plugin

jolestar commented on pkg/server/nicpool.go:
```diff
@@ -0,0 +1,192 @@
+			cachedlist = append(cachedlist, pkg.StringPtr(nic))
+			log.Debugf("Got nic %s in nic pool", nic)
+		}
+		err := pool.nicProvider.ReclaimNic(cachedlist)
```
释放空闲 nic 的操作可以提供个接口，由系统调用。不要在 shutdown pool 的时候进行操作，因为关闭 hostnic daemon 的时候是在 IaaS 操作的一个 job 中，在 job 中发起新的 job 会导致死锁。

jolestar commented on pkg/server/nicpool.go:
```diff
@@ -0,0 +1,192 @@
+	pool := &NicPool{
+		nicDict:               cmap.New(),
+		nicpool:               make(chan string, size),
+		nicReadyPool:          make(chan string, ReadyPoolSize),
```
感觉  nicReadyPool size 可以直接设置成系统支持的最大 nic 数，如果 readypool 放不进去，就说明到达上限了，直接删除就行。然后有个后台的 go routine 定时清理 readypool 到一个合理的值，比如 ReadyPoolSize。 感觉这样逻辑也更清楚些，对突发的大量容器反复创建销毁的支持也更友好些。你考虑下。

Commit:
remove unused import to eliminate import cycle

Commit:
add cleanup cache command

Commit:
update service specs

Commit:
hostnic-cni-18 WIP:Daemon process

add clean/stop command
add prometheus endpoint

Commit:
hostnic-cni-18 WIP:Daemon process

add clean/stop command
add prometheus endpoint

Commit:
hostnic-cni-18 WIP:Daemon
add unit test cases

Commit:
hostnic-cni-18 WIP:Daemon
fix unit test cases

Commit:
hostnic-cni-18 WIP:Daemon
update readme

Commit:
hostnic-cni-18 WIP:Daemon
add test case for clearcache
